ledger-core/virtual/integration/deduplication/constructor_test.go

```diff
@@ -24,118 +24,147 @@ import (
-			firstMessage = false
-
-			server, ctx := utils.NewUninitializedServer(nil, t)
-			defer server.Stop()
-
-			executeDone := server.Journal.WaitStopOf(&execute.SMExecute{}, 2)
-
-			runnerMock := logicless.NewServiceMock(ctx, mc, nil)
-			server.ReplaceRunner(runnerMock)
-			server.Init(ctx)
-
-			var (
-				isolation = contract.ConstructorIsolation()
-				outgoing  = server.BuildRandomOutgoingWithPulse()
-				class     = gen.UniqueGlobalRef()
-			)
-
-			pl := payload.VCallRequest{
-				CallType:       payload.CTConstructor,
-				CallFlags:      payload.BuildCallFlags(isolation.Interference, isolation.State),
-				Callee:         class,
-				CallSiteMethod: "New",
-				CallOutgoing:   outgoing,
-			}
-
-			{
-				requestResult := requestresult.New([]byte("123"), gen.UniqueGlobalRef())
-				requestResult.SetActivate(gen.UniqueGlobalRef(), class, []byte("234"))
-
-				executionMock := runnerMock.AddExecutionMock(outgoing.String())
-				executionMock.AddStart(test.executionFn, &execution.Update{
-					Type:   execution.Done,
-					Result: requestResult,
-				})
-			}
-
-			typedChecker := server.PublisherMock.SetTypedChecker(ctx, mc, server)
-			typedChecker.VCallResult.Set(func(result *payload.VCallResult) bool {
-				if test.sendMessageFn != nil {
-					test.sendMessageFn()
-				}
-				return false
-			})
-
-			{
-				server.SendPayload(ctx, &pl)
-			}
-
-			testutils.WaitSignalsTimed(t, 10*time.Second, synchronizeExecution.Wait())
-
-			{
-				server.SendPayload(ctx, &pl)
-			}
-
-			synchronizeExecution.WakeUp()
-
-			testutils.WaitSignalsTimed(t, 10*time.Second, executeDone)
-			testutils.WaitSignalsTimed(t, 10*time.Second, server.Journal.WaitAllAsyncCallsDone())
-
-			{
-				assert.Equal(t, test.resultCount, typedChecker.VCallResult.Count())
-			}
-
-			mc.Finish()
 
+	{
+		requestResult := requestresult.New([]byte("123"), gen.UniqueGlobalRef())
+		requestResult.SetActivate(gen.UniqueGlobalRef(), class, []byte("234"))
+
+		executionMock := runnerMock.AddExecutionMock(outgoing.String())
+		executionMock.AddStart(executionFn, &execution.Update{
+			Type:   execution.Done,
+			Result: requestResult,
 		})
 	}
+
+	pl := payload.VCallRequest{
+		CallType:       payload.CTConstructor,
+		CallFlags:      payload.BuildCallFlags(isolation.Interference, isolation.State),
+		Callee:         class,
+		CallSiteMethod: "New",
+		CallOutgoing:   outgoing,
+	}
+
+	{
+		// send first call request
+		server.SendPayload(ctx, &pl)
+	}
+
+	// await first SMExecute go to step execute (in this point machine is still not publish result to table in SMObject)
+	testutils.WaitSignalsTimed(t, 10*time.Second, synchronizeExecution.Wait())
+
+	{
+		// send second call request
+		server.SendPayload(ctx, &pl)
+	}
+
+	// second SMExecute should stop in deduplication algorithm and she should not send result because she started during execution first machine
+	testutils.WaitSignalsTimed(t, 10*time.Second, server.Journal.WaitStopOf(&execute.SMExecute{}, 1))
```

Comment:
That part won't work: races, races everywhere.

Please re-read carefully how `WaitStopOf()` works and what's the difference between it and `ReplayAndWaitStopOf()`.

Briefly: you should call `WaitStopOf` before `SendPayload`, not after.

Reply:
Same everywhere, carefully re-read your edits and fix that kind of race everywhere, pls

Reply:
Fixed